awx/main/tasks/receptor.py

```diff
@@ -296,41 +284,45 @@ def _run_internal(self, receptor_ctl):
-        # work_unit_id_assigned event then this case may have occured.
-        self.task.instance.work_unit_id = result['unitid']  # Set work_unit_id in-memory only
-        self.task.instance.log_lifecycle("work_unit_id_received")
-        self.task.update_model(self.task.instance.pk, work_unit_id=result['unitid'])
-        self.task.instance.log_lifecycle("work_unit_id_assigned")
-
-        sockin.close()
-        sockout.close()
-
-        if transmitter_thread.exc:
-            raise transmitter_thread.exc[1].with_traceback(transmitter_thread.exc[2])
-
-        transmitter_thread.join()
+        with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
+            transmitter_future = executor.submit(self.transmit, sockin)
+
+            _kw = {}
+            if self.work_type == 'ansible-runner':
+                _kw['node'] = self.task.instance.execution_node
+                use_stream_tls = get_conn_type(_kw['node'], receptor_ctl).name == "STREAMTLS"
+                _kw['tlsclient'] = get_tls_client(use_stream_tls)
+
+            # submit our work, passing in the right side of our socketpair for reading.
+            result = receptor_ctl.submit_work(
+                worktype=self.work_type, payload=sockout.makefile('rb'), params=self.receptor_params, signwork=self.sign_work, **_kw
+            )
+
+            sockin.close()
+            sockout.close()
+
+            self.unit_id = result['unitid']
+            # Update the job with the work unit in-memory so that the log_lifecycle
+            # will print out the work unit that is to be associated with the job in the database
+            # via the update_model() call.
+            # We want to log the work_unit_id as early as possible. A failure can happen in between
+            # when we start the job in receptor and when we associate the job <-> work_unit_id.
+            # In that case, there will be work running in receptor and Controller will not know
+            # which Job it is associated with.
+            # We do not programatically handle this case. Ideally, we would handle this with a reaper case.
+            # The two distinct job lifecycle log events below allow for us to at least detect when this
+            # edge case occurs. If the lifecycle event work_unit_id_received occurs without the
+            # work_unit_id_assigned event then this case may have occured.
+            self.task.instance.work_unit_id = result['unitid']  # Set work_unit_id in-memory only
+            self.task.instance.log_lifecycle("work_unit_id_received")
+            self.task.update_model(self.task.instance.pk, work_unit_id=result['unitid'])
+            self.task.instance.log_lifecycle("work_unit_id_assigned")
+
+        try:
+            transmitter_future.result()
+        except Exception as exc:
+            logger.exception(f'An error was encountered while transmitting the payload for {self.unit_id}')
```

Comment:
I don't agree with using `logger.exception` if it's just going to raise the exception afterwards, because logs wind up getting duplicated due to the `raise`. I would suggest that we _do not log at all_. In another issue, we are hunting an issue with the setting local cache, because it's not thread-safe.